### ISSUE TITLE
Uncaught exception calling delete method on undefined ipConnections in Connections.onConnectionClosed

### ISSUE THREAD

**Issue by congratulatio** [wight-airmash/ab-server#44]
From airbattle.log:

    {"level":30,"time":1570021705066,"msg":"uncaughtException caught","v":1}
    {"level":50,"time":1570021705066,"msg":"error caused exit","stack":"TypeError: Cannot read property 'delete' of undefined\n    at Connections.onConnectionClosed (/app/dist/server/connections.js:57:27)\n    at EventEmitter.emit (/app/node_modules/eventemitter3/index.js:181:35)\n    at close (/app/dist/endpoints/ws.js:150:33)","type":"Error","v":1}
    {"level":30,"time":1570021705067,"msg":"exit caught","v":1}

Call stack:

    TypeError: Cannot read property 'delete' of undefined
        at Connections.onConnectionClosed (/app/dist/server/connections.js:57:27)
        at EventEmitter.emit (/app/node_modules/eventemitter3/index.js:181:35)
        at close (/app/dist/endpoints/ws.js:150:33)

This corresponds to the following code in the server:

https://github.com/wight-airmash/ab-server/blob/66355bf06384286eed26b8f0197c55acec201e9e/src/server/connections.ts#L97

**Comment by wight-airmash** [comment]
Fixed in v4.7.2.

I will write about what led to the exception when the update will deployed to the servers.

**Comment by congratulatio** [comment]
Thank you for the quick fix, I will let you know when it's deployed on eu.airmash.online/ffa.

**Comment by congratulatio** [comment]
eu.airmash.online/ffa is now on 4.7.2

**Comment by wight-airmash** [comment]
All data related to the player is stored in the storages. The storages are filled step by step as requests are received from the player to the server. After the connection between the player and the server is established, the storages related to the connection are filled (synchronously). Next, the player's storages are expected to be filled.

But, as with many server events, the event of a player's joining the game is delayed (in the channel `CHANNEL_CONNECT_PLAYER`). The player is not joined to the game immediately after receiving the Login packet, but instead a deferred event is created, which is triggered once per tick (i.e. the player's storages are filled asynchronously) by event `PLAYERS_EMIT_CHANNEL_CONNECT` (this is the main loop event that performs deferred events in the channel `CHANNEL_CONNECT_PLAYER`). This event fills almost all of the storages related to the player. The `connectionByIPList` repository refers to the player — this is not obvious and it is my logical mistake, which eventually led to the crash.

Here is a series of events that led to the crash of the server:
1. The client has established a connection with the server.
2. The client sent a Login packet.
3. The server has created a delayed player creation event in the channel `CHANNEL_CONNECT_PLAYER`.
4. The client has broken the connection.
5. The server tried to clear the storages connected with the connection. The server tried to clear the `connectionByIPList` storage, which has not yet been filled. I did not check its availability because I forgot that it is not really a connection, but a player storage.
6. Server crashed.

It is important that all this happened within 16 milliseconds and between two events `PLAYERS_EMIT_CHANNEL_CONNECT`. That's why this crash was only once during the whole time of the server launch and was not caught earlier.
However, it was not a random error and it could be abused intentionally.

@BlackCrawler, please don't forget to deploy an update.
